### PR TITLE
Feature: Support dbt singular tests

### PR DESCRIPTION
**Pull request by crericha**
This PR introduces `StandaloneAudit`, which is a new Snapshot node type that is currently used to support dbt singular tests. A follow-on PR will add `StandaloneAudit` natively to SQLMesh.

The PR also brings in support for dbt's built-in relationship tests.

image

Remaining work to go from draft to PR:
- Add unit tests
- Update comments that assume snapshot == model

### CODE REVIEW

File: sqlmesh/core/audit/definition.py [hunk 1]
```diff
@@ -247,22 +261,147 @@ def macro_definitions(self) -> t.List[d.MacroDef]:
...
+    def sorted_python_env(self) -> t.List[t.Tuple[str, Executable]]:
+        """Returns the python env sorted by executable kind and then var name."""
+        return sorted(self.python_env.items(), key=lambda x: (x[1].kind, x[0]))
+
+    @property
+    def dialect(self) -> str:
+        return self.audit.dialect
+
+    @property
+    def data_hash(self) -> str:
```
Comment (tobymao) [inline sqlmesh/core/audit/definition.py:333]: audits don't have data hashes

File: sqlmesh/core/console.py [hunk 2]
```diff
@@ -416,26 +416,48 @@ def show_model_difference_summary(
...
+            name for name, snapshot in context_diff.added.items() if snapshot.is_model
+        } - ignored_snapshot_names
         if added_model_names:
             added_tree = Tree(f"[bold][added]Added Models:")
             for model_name in added_model_names:
                 added_tree.add(f"[added]{model_name}")
             tree.add(added_tree)
 
-        removed_model_names = context_diff.removed - ignored_snapshot_names
+        added_audit_names = {
```
Comment (tobymao) [inline sqlmesh/core/console.py:428]: why did you add this to the ui? we don't need to show audits in the console diff

File: sqlmesh/core/console.py [hunk 3]
```diff
@@ -1094,20 +1118,42 @@ def show_model_difference_summary(
...
+            name for name, snapshot in context_diff.added.items() if snapshot.is_model
+        } - ignored_snapshot_names
         if added_model_names:
             self._print(f"**Added Models:**\n")
             for model_name in added_model_names:
                 self._print(f"- {model_name}\n")
             self._print("\n")
 
-        removed_model_names = context_diff.removed - ignored_snapshot_names
+        added_audit_names = {
```
Comment (tobymao) [inline sqlmesh/core/console.py:1130]: remove all of this imo

File: sqlmesh/core/context_diff.py [hunk 4]
```diff
@@ -41,9 +41,9 @@ class ContextDiff(PydanticModel):
     """Whether the currently stored environment record is in unfinalized state."""
     create_from: str
     """The name of the environment the target environment will be created from if new."""
-    added: t.Set[str]
```
Comment (tobymao) [inline sqlmesh/core/context_diff.py:44]: does any of this have to change?

File: sqlmesh/core/console.py (hunk 2 above)
Comment (crericha) [reply sqlmesh/core/console.py:428]: Discussed offline

File: sqlmesh/core/context_diff.py (hunk 4 above)
Comment (crericha) [reply sqlmesh/core/context_diff.py:44]: Added the snapshots to the methods, so that consumers can easily interrogate information about the snapshot. In my case, the type of node.

I originally added an added, removed, modified field for each node type, but that wouldn't scale well if more types were added.

File: sqlmesh/core/audit/definition.py [hunk 5]
```diff
@@ -163,7 +172,7 @@ def load_multiple(
 
     def render_query(
         self,
-        snapshot_or_model: t.Union[Snapshot, _Model],
+        snapshot_or_node: t.Union[Snapshot, _Model, StandaloneAudit],
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:175]: Why not `t.Union[Snapshot, Node]` then?

Comment (izeigerman) [reply sqlmesh/core/audit/definition.py:175]: This feels wrong to me. Why do we even need to pass `StandaloneAudit` here if we don't even use it? Should this argument just be Optional?

File: sqlmesh/core/audit/definition.py [hunk 6]
```diff
@@ -247,22 +261,154 @@ def macro_definitions(self) -> t.List[d.MacroDef]:
...
+
+    _path: Path = Path()
+    _depends_on: t.Optional[t.Set[str]] = None
+
+    @property
+    def depends_on(self) -> t.Set[str]:
+        if self._depends_on is None:
+            self._depends_on = self.depends_on_ or set()
+
+            query = exp.select("*").from_("table")  # self.render_query(optimize=False)
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:312]: What is this for? Why are we doing this?

File: sqlmesh/core/audit/definition.py [hunk 7]
```diff
@@ -282,3 +428,7 @@ def _maybe_parse_arg_pair(e: exp.Expression) -> t.Tuple[str, exp.Expression]:
     if isinstance(e, exp.EQ):
         return e.left.name, e.right
     raise_config_error(f"Invalid defaults expression: {e}", error_type=AuditConfigError)
+
+
+def is_audit(node: Node) -> bool:
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:433]: I think this is used in exactly one place. Do we really need a function for this?

File: sqlmesh/core/context_diff.py [hunk 8]
```diff
@@ -41,9 +41,9 @@ class ContextDiff(PydanticModel):
...
+    added: t.Dict[str, Snapshot]
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:44]: This feels redundant. Why users can just use the key with the `snapshots` dict?

File: sqlmesh/core/context_diff.py [hunk 9]
```diff
@@ -41,9 +41,9 @@ class ContextDiff(PydanticModel):
...
     """New models."""
-    removed: t.Set[str]
+    removed: t.Dict[str, Snapshot]
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:46]: Ditto

File: sqlmesh/core/model/definition.py [hunk 10]
```diff
@@ -1916,3 +1917,7 @@ def _refs_to_sql(values: t.Any) -> exp.Expression:
     "hash_raw_query": exp.convert,
     "table_properties_": lambda value: value,
 }
+
+
+def is_model(node: Node) -> bool:
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1922]: Shouldn't this and `is_audit` be a property of `Node`?

File: sqlmesh/core/model/kind.py [hunk 11]
```diff
@@ -88,6 +88,7 @@ class ModelKindName(str, ModelKindMixin, Enum):
     EMBEDDED = "EMBEDDED"
     SEED = "SEED"
     EXTERNAL = "EXTERNAL"
+    NONE = "NONE"
```
Comment (izeigerman) [inline sqlmesh/core/model/kind.py:91]: Where is this used?

File: sqlmesh/core/plan/definition.py [hunk 12]
```diff
@@ -441,7 +442,7 @@ def has_unmodified_unpromoted(self) -> bool:
 
     def _add_restatements(self) -> None:
         def is_restateable_snapshot(snapshot: Snapshot) -> bool:
-            if not self.is_dev and snapshot.model.disable_restatement:
+            if not self.is_dev and snapshot.is_model and snapshot.model.disable_restatement:
```
Comment (izeigerman) [inline sqlmesh/core/plan/definition.py:445]: I feel like `disable_restatement` should be a part of `Snapshot` and look like this:
```
return self.is_model and self.model.disable_restatement
```

File: sqlmesh/core/plan/definition.py [hunk 13]
```diff
@@ -391,7 +392,7 @@ def _set_effective_from(self, effective_from: t.Optional[TimeLike]) -> None:
         self._effective_from = effective_from
 
         for snapshot in self.new_snapshots:
-            if not snapshot.model.disable_restatement:
+            if snapshot.is_model and not snapshot.model.disable_restatement:
```
Comment (izeigerman) [inline sqlmesh/core/plan/definition.py:395]: Ditto about making `disable_restatement` a property of Snapshot

File: sqlmesh/core/scheduler.py [hunk 14]
```diff
@@ -140,7 +140,10 @@ def evaluate(
             snapshot.name: snapshot,
         }
 
-        if isinstance(snapshot.model, SeedModel) and not snapshot.model.is_hydrated:
+        if (
+            isinstance(snapshot.node, SeedModel)
+            and not t.cast(SeedModel, snapshot.node).is_hydrated
```
Comment (izeigerman) [inline sqlmesh/core/scheduler.py:145]: Why can't this remain `snapshot.model.is_hydrated`?

File: sqlmesh/core/snapshot/definition.py [hunk 15]
```diff
@@ -42,7 +42,8 @@
 
 Interval = t.Tuple[int, int]
 Intervals = t.List[Interval]
-SnapshotNode = Annotated[Model, Field(discriminator="source_type")]
+NodeClasses = t.Union[Model, StandaloneAudit]
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:45]: Let's not use `Classes` as part of the type name. To me it feels like it's a list of type **instances** not a type. I suggest perhaps following the `Node` (for union) vs `_Node` (for base class) like we do everywhere else.

Comment (izeigerman) [reply sqlmesh/core/snapshot/definition.py:45]: Seems like it's only used in the next line. Can we just inline?

File: sqlmesh/core/snapshot/evaluator.py [hunk 16]
```diff
@@ -142,6 +141,11 @@ def apply(query_or_df: QueryOrDF, index: int = 0) -> None:
             **common_render_kwargs,
         )
 
+        if not snapshot.is_model:
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:144]: Why wait so long, why not return at the top?

File: sqlmesh/utils/errors.py [hunk 17]
```diff
@@ -56,15 +59,20 @@ class AuditConfigError(ConfigError):
...
-        self, audit_name: str, model_name: str, count: int, query: exp.Subqueryable
+        self,
+        audit_name: str,
+        count: int,
+        query: exp.Subqueryable,
+        model: t.Optional[Model] = None,
     ) -> None:
         self.audit_name = audit_name
-        self.model_name = model_name
+        self.model_name = model.name if model else None
```
Comment (izeigerman) [inline sqlmesh/utils/errors.py:69]: `model_name` can just be a property method.

File: sqlmesh/core/snapshot/evaluator.py [hunk 18]
```diff
@@ -490,7 +484,7 @@ def _audit(
         if audit.skip:
             return AuditResult(
                 audit=audit,
-                model=snapshot.model,
+                model=snapshot.model if snapshot.is_model else None,
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:487]: This seems like a fairly common pattern. Can this be a `snapshot.model_or_none` instead?

File: sqlmesh/core/snapshot/evaluator.py [hunk 19]
```diff
@@ -767,21 +761,25 @@ def demote(
...
 
     def create(
         self,
-        model: Model,
+        snapshot: Snapshot,
         name: str,
         **render_kwargs: t.Any,
     ) -> None:
+        model = snapshot.model
+        self.adapter.execute(model.render_pre_statements(**render_kwargs))
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:781]: Why did we move this for `create` but not for `evaluate`?

Comment (izeigerman) [reply sqlmesh/core/snapshot/evaluator.py:781]: What if `create` is overridden in one of the strategies, do they also need to invoke pre / post statements too?

File: sqlmesh/core/snapshot/evaluator.py [hunk 20]
```diff
@@ -958,10 +964,13 @@ def append(
...
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:972]: That's what I'm talking about. What's the point of moving pre / post statements into strategies?

File: sqlmesh/core/state_sync/cache.py [hunk 21]
```diff
@@ -68,8 +68,8 @@ def get_snapshots(
             elif snapshot:
                 if (
                     hydrate_seeds
-                    and isinstance(snapshot.model, SeedModel)
-                    and not snapshot.model.is_hydrated
+                    and isinstance(snapshot.node, SeedModel)
+                    and not t.cast(SeedModel, snapshot.model).is_hydrated
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/cache.py:72]: Why do we cast in some places but not others?

File: sqlmesh/core/state_sync/cache.py [hunk 22]
```diff
@@ -80,7 +80,7 @@ def get_snapshots(
 
         for snapshot_id, snapshot in existing.items():
             cached = self._from_cache(snapshot_id, now)
-            if cached and (not isinstance(cached.model, SeedModel) or cached.model.is_hydrated):
+            if cached and (not isinstance(cached.node, SeedModel) or cached.node.is_hydrated):
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/cache.py:83]: We don't cast here

**Comment by izeigerman** [comment]
Shall we add a standalone audit to the example dbt project? Or is there already one?

File: sqlmesh/core/audit/definition.py (hunk 5 above)
Comment (crericha) [reply sqlmesh/core/audit/definition.py:175]: The reason it explicitly unions the types is the use of `python_env`, which is not part of the `Node` api and will be needed for SQLMesh standalone tests.

I could change the type to `Node` and check for `StandaloneAudit` or `_Model` in `_create_query_renderer`. That's probably cleaner.

File: sqlmesh/core/audit/definition.py (hunk 6 above)
Comment (crericha) [reply sqlmesh/core/audit/definition.py:312]: Temporary code. Needs to be fixed.

File: sqlmesh/core/audit/definition.py (hunk 7 above)
Comment (crericha) [reply sqlmesh/core/audit/definition.py:433]: I did a grep and found it used 6 times.

File: sqlmesh/core/context_diff.py (hunk 9 above)
Comment (crericha) [reply sqlmesh/core/context_diff.py:46]: If I'm reading the code correctly, removed snapshots are not part of ContextDiff. If these were added to ContextDiff, then yes, the dicts could be sets.

File: sqlmesh/core/model/definition.py (hunk 10 above)
Comment (crericha) [reply sqlmesh/core/model/definition.py:1922]: Is the suggestion to add to `Node` and have it return false, then override in `_Model` and `StandaloneAudit`?

File: sqlmesh/core/model/kind.py (hunk 11 above)
Comment (crericha) [reply sqlmesh/core/model/kind.py:91]: Default value for `SnapshotTableInfo` `kind_name`, so that `model_kind_name` doesn't have to be optional (which had a spider effect throughout a lot of code).

File: sqlmesh/core/scheduler.py (hunk 14 above)
Comment (crericha) [reply sqlmesh/core/scheduler.py:145]: `is_hydrated` is only supported for `SeedModel` and not part of the general `Model` interface, so it needs to be narrowed down. Maybe `snapshot.node.is_hydrated` will work though. I'll try that.

File: sqlmesh/core/snapshot/definition.py [hunk 23]
```diff
@@ -292,7 +310,8 @@ class SnapshotTableInfo(PydanticModel, SnapshotInfoMixin, frozen=True):
     parents: t.Tuple[SnapshotId, ...]
     previous_versions: t.Tuple[SnapshotDataVersion, ...] = ()
     change_category: t.Optional[SnapshotChangeCategory] = None
-    kind_name: ModelKindName
+    kind_name: ModelKindName = ModelKindName.NONE
```
[the remark this answers is not in the document]
Comment (crericha) [reply sqlmesh/core/snapshot/definition.py:313]: I tried to make it Optional at first, but it spiraled out of control as I updated clients of `SnapshotTableInfo` to support an Optional value for `model_kind_name`. IMHO, this approach made the code more readable.

File: sqlmesh/core/snapshot/evaluator.py (hunk 16 above)
Comment (crericha) [reply sqlmesh/core/snapshot/evaluator.py:144]: I can't recall why. Will move to the top.

File: sqlmesh/core/snapshot/evaluator.py (hunk 18 above)
Comment (crericha) [reply sqlmesh/core/snapshot/evaluator.py:487]: I like that!

File: sqlmesh/core/snapshot/evaluator.py (hunk 19 above)
Comment (crericha) [reply sqlmesh/core/snapshot/evaluator.py:781]: I'll move it back for `create` and early return if not a `model` like done in `evaluate`.

**Comment by crericha** [comment]
> Shall we add a standalone audit to the example dbt project? Or is there already one?

There is one already.

File: sqlmesh/core/context_diff.py (hunk 9 above)
Comment (izeigerman) [reply sqlmesh/core/context_diff.py:46]: I see. Do we actually need a full-blown snapshot or the Table Info from the environment will suffice?

File: sqlmesh/core/model/definition.py (hunk 10 above)
Comment (izeigerman) [reply sqlmesh/core/model/definition.py:1922]: yeah, that would work too

File: sqlmesh/core/model/kind.py (hunk 11 above)
Comment (izeigerman) [reply sqlmesh/core/model/kind.py:91]: It did? I thought the impact was minimal since we did `==` everywhere. I could only find like 2 places where it matters. Did you find more?

File: sqlmesh/core/context_diff.py (hunk 9 above)
Comment (crericha) [reply sqlmesh/core/context_diff.py:46]: The TableInfo would suffice. We have the removed Snapshot in a local var when constructing context diff. I can add a removed_snapshots dict if that's preferable. What do you think?

File: sqlmesh/core/model/kind.py (hunk 11 above)
Comment (crericha) [reply sqlmesh/core/model/kind.py:91]: The method originates from the `ModelKindMixin` and making the return type optional there is what caused the issue.

File: sqlmesh/core/audit/definition.py [hunk 24]
```diff
@@ -247,22 +261,173 @@ def macro_definitions(self) -> t.List[d.MacroDef]:
...
+    python_env_: t.Optional[t.Dict[str, Executable]] = Field(default=None, alias="python_env")
+    source_type: Literal["audit"] = "audit"
+
+    _path: Path = Path()
+    _depends_on: t.Optional[t.Set[str]] = None
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_fields(
+        cls, values: t.Dict[str, t.Union[t.Tuple[str, ...], t.Optional[str], t.Dict[str, t.Any]]]
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:317]: I think this type signature is useless to the point that you're better off just using `t.Dict[str, t.Any]`. Nobody can decipher this anyway. Same for return value.

File: sqlmesh/core/audit/definition.py [hunk 25]
```diff
@@ -247,22 +261,173 @@ def macro_definitions(self) -> t.List[d.MacroDef]:
...
+    @property
+    def data_hash(self) -> str:
+        """
+        Computes the data hash for the node.
+
+        Returns:
+            The data hash for the node.
+        """
+        # StandaloneAudits do not have a data hash
+        return ""
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:357]: This should be either `"0"` or `hash_data("")`.

File: sqlmesh/core/console.py [hunk 26]
```diff
@@ -483,6 +458,75 @@ def _get_ignored_tree(
             )
         return ignored
 
+    def _show_summary_tree_for(
+        self,
+        context_diff: ContextDiff,
+        header: str,
+        filter: t.Callable,
```
Comment (izeigerman) [inline sqlmesh/core/console.py:465]: `filter` is an existing Python keyword (or rather builtin function). I suggest coming up with a different name.

Comment (izeigerman) [reply sqlmesh/core/console.py:465]: If it's a predicate than I think expanding the type signature makes sense: `t.Callable[[Snapshot], bool]`.

File: sqlmesh/core/console.py [hunk 27]
```diff
@@ -483,6 +458,75 @@ def _get_ignored_tree(
...
+        removed_names = {
+            name for name, snapshot in context_diff.removed_snapshots.items() if filter(snapshot)
+        } - filtered_ignored_names
+        modified_names = {
+            name
+            for name, snapshots in context_diff.modified_snapshots.items()
+            if filter(snapshots[0])
+        } - filtered_ignored_names
+
+        if (
```
Comment (izeigerman) [inline sqlmesh/core/console.py:486]: [Nit] 
```
name_sets = (added_names, removed_names, modified_names, filtered_ignored_names)
if all(not names for names in name_sets):
    return
```

File: sqlmesh/core/context_diff.py [hunk 28]
```diff
@@ -172,7 +177,9 @@ def create(
             is_unfinalized_environment=bool(env and not env.finalized_ts),
             create_from=create_from,
             added=added,
-            removed=removed,
+            removed_snapshots={
+                name: info for name, info in existing_info.items() if name in removed
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:181]: [Nit] Considering that we expect `removed` to be empty most of the time I think you're better off iterating over `removed` instead and then getting the `info` instance from the `existing_info` dict.

File: sqlmesh/core/context_diff.py [hunk 29]
```diff
@@ -125,8 +127,11 @@ def create(
             modified = modified_info.get(name)
             existing = stored.get(snapshot.snapshot_id)
 
-            if existing:
-                # Keep the original model instance to preserve the query cache.
+            if modified and snapshot.node_type != modified.node_type:
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:130]: Shouldn't we remove it from the `modified` mapping then?

File: sqlmesh/core/audit/definition.py (hunk 25 above)
Comment (crericha) [reply sqlmesh/core/audit/definition.py:357]: The method returns a string. I used `""` to indicate there is no data hash vs a static data dash that `"0" `or `hash_data("") `would produce. Are you thinking there is no value in indicating there isn't a data hash?

File: sqlmesh/core/context_diff.py [hunk 30]
```diff
@@ -202,77 +214,81 @@ def promotable_models(self) -> t.Set[str]:
...
-        if model not in self.modified_snapshots:
+        if node not in self.snapshots:
+            raise SQLMeshError(f"`{node}` does not exist.")
+        if node not in self.modified_snapshots:
             return ""
 
-        new, old = self.modified_snapshots[model]
-        return old.model.text_diff(new.model)
+        new, old = self.modified_snapshots[node]
+        if new.is_model:
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:290]: Shouldn't `text_diff` be a behavior of a snapshot or a node to avoid this branching?

File: sqlmesh/core/audit/__init__.py [hunk 31]
```diff
@@ -3,7 +3,14 @@
...
 from sqlmesh.core.audit import builtin
-from sqlmesh.core.audit.definition import Audit, AuditResult
+from sqlmesh.core.audit.definition import (
+    Audit,
+    AuditResult,
+    StandaloneAudit,
+    create_standalone_audit,
+)
+
+AuditType = t.Union[Audit, StandaloneAudit]
```
Comment (izeigerman) [inline sqlmesh/core/audit/__init__.py:13]: I don't think a type contain `Type` in its name.

File: sqlmesh/core/context_diff.py (hunk 29 above)
Comment (crericha) [reply sqlmesh/core/context_diff.py:130]: modified isn't used again, but I can remove it to prevent bugs in future modifications.

File: sqlmesh/core/model/kind.py [hunk 32]
```diff
@@ -88,17 +88,18 @@ class ModelKindName(str, ModelKindMixin, Enum):
...
```
Comment (izeigerman) [inline sqlmesh/core/model/kind.py:91]: Why is the value still here?

File: sqlmesh/core/snapshot/definition.py [hunk 33]
```diff
@@ -291,7 +304,8 @@ class SnapshotTableInfo(PydanticModel, SnapshotInfoMixin, frozen=True):
     parents: t.Tuple[SnapshotId, ...]
     previous_versions: t.Tuple[SnapshotDataVersion, ...] = ()
     change_category: t.Optional[SnapshotChangeCategory] = None
-    kind_name: ModelKindName
+    kind_name: t.Optional[ModelKindName] = None
+    node_type_: NodeType = Field(default=NodeType.MODEL, alias="node_type")
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:308]: Why do we need an alias if the property method just returns the value as is?

Comment (izeigerman) [reply sqlmesh/core/snapshot/definition.py:308]: Why is this a property of the Snapshot and not the `_Node`?

File: sqlmesh/core/snapshot/evaluator.py [hunk 34]
```diff
@@ -978,10 +987,13 @@ def append(
...
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/evaluator.py:995]: Why is this still here?

File: sqlmesh/core/model/kind.py (hunk 32 above)
Comment (crericha) [reply sqlmesh/core/model/kind.py:91]: Not needed. Will remove.

File: sqlmesh/core/audit/__init__.py (hunk 31 above)
Comment (crericha) [reply sqlmesh/core/audit/__init__.py:13]: Habit from that being the coding standard from where I worked.

File: sqlmesh/core/snapshot/definition.py (hunk 33 above)
Comment (crericha) [reply sqlmesh/core/snapshot/definition.py:308]: Its a property of SnapshotTableInfo, which doesn't have the node to figure out the type.

File: sqlmesh/core/snapshot/evaluator.py (hunk 34 above)
Comment (crericha) [reply sqlmesh/core/snapshot/evaluator.py:995]: Will fix.

**Comment by izeigerman** [comment]
Please make sure that Airflow tests pass locally before merging

File: sqlmesh/core/audit/definition.py [hunk 35]
```diff
@@ -12,66 +13,215 @@
...
-    dialect: str = ""
-    """The dialect of the audit query."""
-    skip: bool = False
-    """Setting this to `true` will cause this audit to be skipped. Defaults to `false`."""
-    blocking: bool = True
-    """Setting this to `true` will cause the pipeline execution to stop if this audit fails. Defaults to `true`."""
-    defaults: t.Dict[str, exp.Expression] = {}
-    """Default values for the audit query."""
+    dialect: str
+    skip: bool
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:56]: Why should this be a part of the mixin?

File: sqlmesh/core/audit/definition.py [hunk 36]
```diff
@@ -12,66 +13,215 @@
...
+    @property
+    def macro_definitions(self) -> t.List[d.MacroDef]:
+        """All macro definitions from the list of expressions."""
+        return [s for s in self.expressions if isinstance(s, d.MacroDef)]
+
+    def _create_query_renderer(self, node: _Node) -> QueryRenderer:
+        kwargs: t.Dict[str, t.Any] = {}
+
+        node_dialect = ""
+        if node.is_model:
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:160]: Now when you use inheritence, can you just override and avoid branching here?

File: sqlmesh/core/audit/definition.py [hunk 37]
```diff
@@ -12,66 +13,215 @@
...
+        Returns:
+            The rendered expression.
+        """
+        from sqlmesh.core.snapshot import Snapshot
+
+        node = snapshot_or_node if isinstance(snapshot_or_node, _Node) else snapshot_or_node.node
+        query_renderer = self._create_query_renderer(node)
+        extra_kwargs = {}
+
+        if node.is_model:
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:98]: Can we just override the method now and avoid doing these checks?

File: sqlmesh/core/audit/definition.py [hunk 38]
```diff
@@ -287,45 +179,184 @@ def _create_query_renderer(self, node: _Node) -> QueryRenderer:
         )
 
 
-class StandaloneAudit(_Node):
+@field_validator("name", "dialect", mode="before", check_fields=False)
+def audit_string_validator(v: t.Any) -> t.Optional[str]:
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:183]: Shouldn't this be a class method and take class as the first argument?

File: sqlmesh/core/audit/definition.py [hunk 39]
```diff
@@ -287,45 +179,184 @@ def _create_query_renderer(self, node: _Node) -> QueryRenderer:
...
+    if isinstance(v, exp.Expression):
+        return v.name.lower()
+    return str(v).lower() if v is not None else None
+
+
+@field_validator("defaults", mode="before", check_fields=False)
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:189]: ditto

File: sqlmesh/core/node.py [hunk 40]
```diff
@@ -299,6 +301,17 @@ def cron_floor(self, value: TimeLike) -> TimeLike:
         """
         return self.croniter(self.cron_next(value)).get_prev()
 
+    def text_diff(self, other: _Node) -> str:
```
Comment (izeigerman) [inline sqlmesh/core/node.py:304]: This should be just `Node`.

File: sqlmesh/core/snapshot/definition.py [hunk 41]
```diff
@@ -510,7 +510,7 @@ def from_node(
         node_audits = (
             tuple(t.cast(_Model, node).referenced_audits(audits or {}))
             if node.is_model
-            else tuple([t.cast(StandaloneAudit, node).audit])
+            else tuple()
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:513]: Shouldn't this be `tuple(node)`?

File: sqlmesh/dbt/test.py [hunk 42]
```diff
@@ -121,27 +121,32 @@ def to_sqlmesh(self, context: DbtContext) -> AuditType:
...
+                name=self.name,
+                dialect=context.dialect,
+                skip=skip,
+                query=query,
+                jinja_macros=jinja_macros,
+                depends_on={model.sql_name for model in test_context.refs.values()}.union(
+                    {source.sql_name for source in test_context.sources.values()}
+                ),
+                tags=self.tags,
+                # TODO node values
```
Comment (izeigerman) [inline sqlmesh/dbt/test.py:136]: what's node values?

File: sqlmesh/core/audit/definition.py [hunk 43]
```diff
@@ -374,54 +401,34 @@ def metadata_hash(self, audits: t.Dict[str, Audit]) -> str:
...
-            else self.audit.render_query(self) or self.audit.query
-        )
+        query = self.query if self.hash_raw_query else self.render_query(self) or self.query
         data.append(query.sql(comments=False))
 
         return hash_data(data)
 
-    def text_diff(self, other: StandaloneAudit) -> str:
-        """Produce a text diff against another standalone audit.
+    def text_diff(self, other: _Node) -> str:
```
Comment (izeigerman) [inline sqlmesh/core/audit/definition.py:409]: Ditto: this should just be `Node`.

File: sqlmesh/core/model/definition.py [hunk 44]
```diff
@@ -411,15 +412,20 @@ def referenced_audits(self, audits: t.Dict[str, Audit]) -> t.List[Audit]:
                 )
         return referenced_audits
 
-    def text_diff(self, other: Model) -> str:
-        """Produce a text diff against another model.
+    def text_diff(self, other: _Node) -> str:
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:415]: Ditto

File: sqlmesh/core/model/definition.py [hunk 45]
```diff
@@ -1179,7 +1185,7 @@ def render(
             df[string_columns] = df[string_columns].astype(str)
             yield df
 
-    def text_diff(self, other: Model) -> str:
+    def text_diff(self, other: _Node) -> str:
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1188]: Ditto

File: sqlmesh/core/snapshot/definition.py (hunk 41 above)
Comment (crericha) [reply sqlmesh/core/snapshot/definition.py:513]: This is for attached audits. I can maybe change the name.

File: sqlmesh/dbt/test.py (hunk 42 above)
Comment (crericha) [reply sqlmesh/dbt/test.py:136]: Node class fields. I plan to add those in the follow-on PR.

File: sqlmesh/core/snapshot/definition.py (hunk 41 above)
Comment (crericha) [reply sqlmesh/core/snapshot/definition.py:513]: Refactored to a kwarg, so now only used if the node is a model.

File: sqlmesh/core/plan/definition.py [hunk 46]
```diff
@@ -553,7 +553,7 @@ def _categorize_snapshots(self) -> None:
             snapshot = self._snapshot_mapping.get(name)
             if not snapshot or snapshot.change_category:
                 continue
-            upstream_model_names = self._dag.upstream(name)
+            self._dag.upstream(name)
```
Comment (izeigerman) [inline sqlmesh/core/plan/definition.py:556]: This is a dead line

**Review by izeigerman — APPROVED** [review]
One last comment, LGTM otherwise.
